robolectric/src/main/java/org/robolectric/Robolectric.java

```diff
@@ -378,4 +385,84 @@ public static Scheduler getBackgroundThreadScheduler() {
   public static void flushBackgroundThreadScheduler() {
     getBackgroundThreadScheduler().advanceToLastPostedRunnable();
   }
+
+  @SuppressWarnings({"NewApi", "unchecked"})
+  private static <T extends Service> T instantiateService(Class<T> serviceClass, Intent intent) {
+    if (RuntimeEnvironment.getApiLevel() >= P) {
+      final LoadedApk loadedApk = getLoadedApk();
+      AppComponentFactory factory = getAppComponentFactory(loadedApk);
+      if (factory != null) {
+        try {
+          return (T)
+              factory.instantiateService(
+                  loadedApk.getClassLoader(), serviceClass.getName(), intent);
+        } catch (ReflectiveOperationException e) {
+          Logger.debug("Failed to instantiate Service using AppComponentFactory", e);
+        } catch (ClassCastException ignored) {
+          // androidx.core.app.CoreComponentFactory can create CompatWrapper instance that may
+          // differ from the requested class.
+        }
+      }
+    }
+    return ReflectionHelpers.callConstructor(serviceClass);
+  }
+
+  @SuppressWarnings({"NewApi", "unchecked"})
+  private static <T extends ContentProvider> T instantiateContentProvider(Class<T> providerClass) {
+    if (RuntimeEnvironment.getApiLevel() >= P) {
+      final LoadedApk loadedApk = getLoadedApk();
+      AppComponentFactory factory = getAppComponentFactory(loadedApk);
+      if (factory != null) {
+        try {
+          return (T)
+              factory.instantiateProvider(loadedApk.getClassLoader(), providerClass.getName());
+        } catch (ReflectiveOperationException e) {
+          Logger.debug("Failed to instantiate ContentProvider using AppComponentFactory", e);
+        } catch (ClassCastException ignored) {
+          // androidx.core.app.CoreComponentFactory can create CompatWrapper instance that may
+          // differ from the requested class.
+        }
+      }
+    }
+    return ReflectionHelpers.callConstructor(providerClass);
+  }
+
+  @SuppressWarnings({"NewApi", "unchecked"})
+  private static <T extends Activity> T instantiateActivity(Class<T> activityClass, Intent intent) {
+    if (RuntimeEnvironment.getApiLevel() >= P) {
+      final LoadedApk loadedApk = getLoadedApk();
+      AppComponentFactory factory = getAppComponentFactory(loadedApk);
+      if (factory != null) {
+        try {
+          return (T)
+              factory.instantiateActivity(
+                  loadedApk.getClassLoader(), activityClass.getName(), intent);
+        } catch (ReflectiveOperationException e) {
+          Logger.debug("Failed to instantiate Activity using AppComponentFactory", e);
+        } catch (ClassCastException ignored) {
+          // androidx.core.app.CoreComponentFactory can create CompatWrapper instance that may
```

Comment:
@utzcoz that's true, but the problem is that if an app has a subclass of SliceProvider:

`public class MySliceProvider extends androidx.slice.SliceProvider {...}`

And then in their test code, they have:

`MySliceProvider mySliceProvider = Robolectric.setupContentProvider(MySliceProvider.class)`

With this PR, before the try/catch was added, `Robolectric.setupContentProvider` will attempt to use the AndroidX CoreComponentFactory, which does not return a MySliceProvider anymore, which results in a ClassCastException

The test code could be:
`SliceProvider mySliceProvider = Robolectric.setupContentProvider(MySliceProvider.class)`

Or

`ContentProvider mySliceProvider = Robolectric.setupContentProvider(MySliceProvider.class)`

However, then tests do not have visibility to methods on `MySliceProvider`, because it is wrapped.

Reply:
I would personally have a preference for `instanceOf` checks vs try/catch, the instanceOf checks seem lighter weight, but that is just a personal preference.

Reply:
Actually, it could be possible to use `isAssignableFrom`:
```
        try {
          ContentProvider instance =
              factory.instantiateProvider(loadedApk.getClassLoader(), providerClass.getName());
          if (instance != null && providerClass.isAssignableFrom(instance.getClass())) {
            return (T) instance;
          }
        } catch (ReflectiveOperationException e) {
          Logger.debug("Failed to instantiate ContentProvider using AppComponentFactory", e);
        }
```

Then you are not potentially catching `ClassCastException` in app code.

Reply:
Thank you guys, fixed as you suggested